Riot/Modules/Room/MXKRoomViewController.m

```diff
@@ -2517,6 +2518,126 @@ - (void)updateCurrentEventIdAtTableBottom:(BOOL)acknowledge
+        if (!currentReadMarkerEvent)
+        {
+            currentReadMarkerEvent = [roomDataSource eventWithEventId:roomDataSource.room.accountData.readMarkerEventId];
+        }
+        
+        // Update the read marker only if the current event is available, and the new event is posterior to it.
+        return currentReadMarkerEvent && (currentReadMarkerEvent.originServerTs <= event.originServerTs);
+    }
+    return YES;
+}
+
+/// Try to update the read marker by looking for an eligible event displayed at the bottom of the tableview
+- (void)updateReadMarkerEvent
+{
+    // Compute the content offset corresponding to the line displayed at the table bottom (just above the toolbar).
+    CGFloat contentBottomOffsetY = _bubblesTableView.contentOffset.y + (_bubblesTableView.frame.size.height - _bubblesTableView.adjustedContentInset.bottom);
+    if (contentBottomOffsetY > _bubblesTableView.contentSize.height)
+    {
+        contentBottomOffsetY = _bubblesTableView.contentSize.height;
+    }
+    // Be a bit less retrictive, consider visible an event at the bottom even if is partially hidden.
+    contentBottomOffsetY += 8;
+    
+    // Consider the visible cells (starting by those displayed at the bottom)
+    NSArray *visibleCells = [_bubblesTableView visibleCells];
+    NSInteger index = visibleCells.count;
+    UITableViewCell *cell;
+    while (index--)
+    {
+        cell = visibleCells[index];
+        
+        // Check whether the cell is actually visible
+        if (cell && (cell.frame.origin.y < contentBottomOffsetY))
+        {
+            if (![cell isKindOfClass:MXKTableViewCell.class])
+            {
+                continue;
+            }
+            
+            MXKCellData *cellData = ((MXKTableViewCell *)cell).mxkCellData;
+            
+            // Only 'MXKRoomBubbleCellData' is supported here for the moment.
+            if (![cellData isKindOfClass:MXKRoomBubbleCellData.class])
+            {
+                continue;
+            }
+
+            MXKRoomBubbleCellData *bubbleData = (MXKRoomBubbleCellData*)cellData;
+            
+            // Prevent to place the read marker on a collapsed cell
+            if (bubbleData.collapsed)
+            {
+                continue;
+            }
+            
+            // Check which bubble component is displayed at the bottom.
+            // For that update each component position.
+            [bubbleData prepareBubbleComponentsPosition];
+            
+            NSArray *bubbleComponents = bubbleData.bubbleComponents;
+            NSInteger componentIndex = bubbleComponents.count;
+            
+            CGFloat bottomPositionY = cell.frame.size.height;
+            
+            MXKRoomBubbleComponent *component;
+            
+            while (componentIndex --)
+            {
+                component = bubbleComponents[componentIndex];
+                if (![cell isKindOfClass:MXKRoomBubbleTableViewCell.class])
```

Comment:
Can we combine this check with the previous `isKindOfClass:MXKTableViewCell.class ` and maybe use the casted version from then on? 🤔

Reply:
Done, it should be much easier to understand now.